### ISSUE TITLE
ERROR: unknown entry type 0xc0/0xc1

### ISSUE THREAD

**Issue by m4rkusxxl** [relan/exfat#81]
Some exfat sticks never get mounted again when they were used with a current windows (10).

Looking up the entry types with google point to the "Stream Extension".

Wouldnt it be better to mount the fs read-only when an unknown type occurs? (So we can at least access the supported data.)

**Comment by relan** [comment]
What's your fuse-exfat version?

Could you share an image with such issue if you can reproduce it?

You can mount a volume read-only from command line, but this won't help in this case because directory _reading_ fails.

**Comment by m4rkusxxl** [comment]
fuse-exfat is version 1.2.7.
I already patched it to skip those entries. (So I was able to recover the files from it.)

I will see, if I can reproduce it. (I personally dont use windows.)

Yes, I know that I can mount volumes read-only. But that would not help, as you already pointed out. What I meant is: Better "serve a fallback read-only mount" than just "print error and bail out".

**Comment by m4rkusxxl** [comment]
A simple patch to just ignore the unknown entry types if the filesystem is mounted ro.
[ignore-unkown-entry-type-if-mounted-ro.patch.gz](https://github.com/relan/exfat/files/1584821/ignore-unkown-entry-type-if-mounted-ro.patch.gz)
(To allow recovery of files.)

**Comment by m4rkusxxl** [comment]
Because upload seems to silently fail and highlighting breaks everything:
https://pastebin.com/FhE4RzT8

**Comment by relan** [comment]
Actually, the version of `fsck` from `repair` branch can fix such FS errors.

**Comment by m4rkusxxl** [comment]
"Stream Extension" entries are no FS errors.
Although I dont know what exact data will be saved there, they will be removed and lost after a repair?

**Comment by relan** [comment]
> "Stream Extension" entries are no FS errors.

Not sure what you mean by "Stream Extensions", but 0x85, 0xc0 and 0xc1 are mandatory entries describing each file. Any entry that is valid per se becomes an error when it's unexpected.

> they will be removed and lost after a repair?

Yes.

**Comment by m4rkusxxl** [comment]
Ah okay. Sorry, i misunderstood something. :)

I formated the stick and copied a few files on it.
In windows I noticed that one was missing.
Back in linux I could see that windows created the "System Volume Information" and the "$RECYCLE.BIN" folder. But the one file was still missing.

I now checked the filesystem in hex viewer:
first entry 0x03 => no label
second 0x81 bitmap
third 0x82 upcase
fourth 0x85 + 0xc0 + 2 times 0xc1 => SysVolInfo (wasnt there when I formated it.)
fifth 0x85 + 0xc0 + 0xc1 => recycle (Wasnt there when I formated it.)
And now the problem, a 0xc0 + 0xc1 of the missing file.

So it seems as if windows moved the entries to put its system entries in the beginning. But it missed by one, overwriting the 0x85 entry.

Ever noticed something similar?

**Comment by relan** [comment]
Your analysis seems correct. What did those orphaned (0xc0, 0xc1) entries contain?

> Ever noticed something similar?

Nope.

**Comment by m4rkusxxl** [comment]
0xc0 contained file infos: a normal file, name length, name hash (non zero, but havent checked the value itself), size/validsize and start cluster look correct.
The 0xc1 entries contain the full and correct name.

**Comment by m4rkusxxl** [comment]
Okay. I think I found something:
Fresh formated drive.
Create files with long filenames.
It will not be placed directly behind 0x82.
There will be some 0x00 entries left before the actual files.
This seems to be problematic in windows.

**Comment by m4rkusxxl** [comment]
The problem is find_slot in node.c:
We first look for a space by bitmap.
Then running check_slot for bitmap/upcasetable/label. It will fail at first.
But then we check the first entry after previously selected block.

Something this should solve it:
```
diff -Nur fuse-exfat-1.2.7.orig/libexfat/node.c fuse-exfat-1.2.7/libexfat/node.c
--- fuse-exfat-1.2.7.orig/libexfat/node.c       2017-05-30 17:18:46.000000000 +0200
+++ fuse-exfat-1.2.7/libexfat/node.c    2017-12-27 23:22:23.000000000 +0100
@@ -856,6 +856,7 @@
                                        return -EIO;
                                case -EINVAL:
                                        /* slot is occupied, continue searching */
+                                       i -= contiguous - 1;
                                        contiguous = 0;
                                        break;
                                }
```

**Comment by m4rkusxxl** [comment]
Three test images:
http://www93.zippyshare.com/v/oXAZxMeu/file.html
image-formated-and-created-files.img: Fuse formated and created four files.
image-after-windows-opened-it.img: Tried to open with windows. No files displayed.
image-after-windows-chkdsk.img: After windows checked and repaired it.

**Comment by m4rkusxxl** [comment]
How to continue with this?

**Comment by relan** [comment]
You are right, there is a bug in `find_slot()`. I'll have some time after the New Year to fix it and release an update.

**Comment by m4rkusxxl** [comment]
Do you need any testing of the patch?

**Comment by relan** [comment]
Sorry for keeping silice for so long, very busy those days.

You patch looks like the easiest solution, I've pushed it as 21fab6b208781d9bc6627258a32f50933d1cd901. It passed all my tests, but I need to add a new one to catch valid entries after 0x00.

**Comment by relan** [comment]
Fix released in v1.2.8. Thanks!

**Comment by anton-veretenenko** [comment]
Hi. I am getting this error "ERROR: unexpected entry type 0xc0 after 0xc0 at 2/3." with mount or fsck calls. 32Gb microsd was used in Android phone and it works fine there, just can not mount it in Ubuntu anymore. Is it related? Version 1.2.8.
